### ISSUE TITLE
Request: Shipping extra information instead of address when needs_shipping_address=false

### ISSUE THREAD

**Issue by webdados** [woocommerce/woocommerce#19258]
On customer emails, my account > orders and on the order preview modal window on wp-admin, the shipping address is not shown if `$order->needs_shipping_address()` returns false.

This can happen because the chosen shipping method does not need a shipping address, so that method will add itself to the array returned by the `woocommerce_order_hide_shipping_address` filter or return false to the `woocommerce_order_needs_shipping_address` filter.

This is great, but there's something missing, IMHO.

Whenever the shipping address is not shown because `$order->needs_shipping_address()` returned false, there should be an action for plugins to inject other details instead of the address.

Example: I've developed a shipping method that consists on the customer to chose a pickup location from a list on the checkout. This method hooks into `woocommerce_order_hide_shipping_address` or `woocommerce_order_needs_shipping_address` in order to hide the shipping address but it needs to show the pickup location details instead of it.

So I suggest that on `templates/emails/email-addresses.php`, `templates/emails/plain/email-addresses.php`, `templates/order/order-details-customer.php` and `includes/admin/list-tables/class-wc-admin-list-table-orders.php` there's an action on the place where the shipping address would be so plugins are able to hook into it and provide usefull information for the customer and shop owner.

**Comment by WPprodigy** [comment]
Not opposed to new hooks, but might be worth seeing if these hooks do the trick for you:

`woocommerce_email_customer_details`
`woocommerce_order_details_after_order_table`
`woocommerce_admin_order_preview_end`

**Comment by webdados** [comment]
`woocommerce_email_customer_details`, `woocommerce_order_details_after_order_table` and `woocommerce_admin_order_preview_end` are after the addresses and not on the blank space left by the absence of the shipping address.

Sure we can use them but visually doesn't make a lot of sense. The reason any shipping plugin would use the filters to remove the shipping address from the emails and order details is that that specific delivery will not happen on a regular address, but rather be a pickup location or something similar, so it makes sense that we can use the same area to inject additional information.

**Comment by webdados** [comment]
Actually `woocommerce_order_details_after_order_table` is before the customer details / addresses and not after.

**Comment by WPprodigy** [comment]
Was just thinking they are, markup wise, right next to the address output. And since you'll likely need some custom CSS no matter what, it could be done.

**Comment by webdados** [comment]
For the `templates/order/order-details-customer.php` this could be something like this:

```
if ( ! defined( 'ABSPATH' ) ) {
	exit;
}
$show_shipping = ! wc_ship_to_billing_address_only() && $order->needs_shipping_address();
if ( !$show_shipping ) $shipping_alternate_info = apply_filters( 'woocommerce_shipping_alternate_info', false, $order );
?>
<section class="woocommerce-customer-details">

	<?php if ( $show_shipping || $shipping_alternate_info ) : ?>

	<section class="woocommerce-columns woocommerce-columns--2 woocommerce-columns--addresses col2-set addresses">
		<div class="woocommerce-column woocommerce-column--1 woocommerce-column--billing-address col-1">

	<?php endif; ?>

	<h2 class="woocommerce-column__title"><?php _e( 'Billing address', 'woocommerce' ); ?></h2>

	<address>
		<?php echo wp_kses_post( $order->get_formatted_billing_address( __( 'N/A', 'woocommerce' ) ) ); ?>

		<?php if ( $order->get_billing_phone() ) : ?>
			<p class="woocommerce-customer-details--phone"><?php echo esc_html( $order->get_billing_phone() ); ?></p>
		<?php endif; ?>

		<?php if ( $order->get_billing_email() ) : ?>
			<p class="woocommerce-customer-details--email"><?php echo esc_html( $order->get_billing_email() ); ?></p>
		<?php endif; ?>
	</address>

	<?php if ( $show_shipping || $shipping_alternate_info ) : ?>

		</div><!-- /.col-1 -->

		<div class="woocommerce-column woocommerce-column--2 woocommerce-column--shipping-address col-2">
			<?php if ( $show_shipping ) { ?>
				<h2 class="woocommerce-column__title"><?php _e( 'Shipping address', 'woocommerce' ); ?></h2>
				<address>
					<?php echo wp_kses_post( $order->get_formatted_shipping_address( __( 'N/A', 'woocommerce' ) ) ); ?>
				</address>
			<?php } else { echo $shipping_alternate_info; } ?>
		</div><!-- /.col-2 -->

	</section><!-- /.col2-set -->

	<?php endif; ?>

</section>
```

And then developers could inject the content there using the filter:

```
add_filter( 'woocommerce_shipping_alternate_info', 'my_woocommerce_shipping_alternate_info', 10, 2 );
function my_woocommerce_shipping_alternate_info( $info, $order ) {
	//Logic would be implemented here, to only return html if needed, according to the order shipping method or whatever...
	ob_start();
	?>
	<h2 class="woocommerce-column__title">Shipping alternate info</h2>
	<p>This is some information to show instead of the shipping address</p>
	<?php
	return ob_get_clean();
}
```

This is just a proof of concept, to exemplify what can be done, and my code does not follow any existent WooCommerce standards or rules. The same concept would be applied to `templates/emails/email-addresses.php`, `templates/emails/plain/email-addresses.php` and `includes/admin/list-tables/class-wc-admin-list-table-orders.php`.

**Comment by mikejolley** [comment]
Adding a niche filter named `woocommerce_shipping_alternate_info` seems very wrong considering the order does really need/have shipping, it's just a custom method which hacks this output by removing setting needs_shipping to false. 

Not saying what you've done is wrong, but it's not supported by core. So having a filter to support this partially/workaround this issue as a result of your change seems wrong. Does that make sense?

If there is no shipping address, the address should not be shown. This area is specifically for shipping addresses. Let's not change that. Add your extra information elsewhere. 

`woocommerce_order_details_after_order_table` seems like a logical place and requires no template modification/theme compatibility.

**Comment by webdados** [comment]
> Adding a niche filter named woocommerce_shipping_alternate_info seems very wrong considering the order does really need/have shipping

Yes, the order does need shipping, but the shipping will not be done to the traditional WooCommerce Shipping Address, so we need to remove it or else the customer will just be confused. That's why I always refer to `$order->needs_shipping_address()`.

The order needs shipping but does not need a shipping address because the shipping will be done by an alternate method.

As I said before:
> The reason any shipping plugin would use the filters to remove the shipping address from the emails and order details is that that specific delivery will not happen on a regular address, but rather be a pickup location or something similar, so it makes sense that we can use the same area to inject additional information.

Because the order is going to be shipped, but not to the Shipping Address, and we removed it via `woocommerce_order_needs_shipping_address`, it does make sense to me that we should have the ability to use that area to give the customer information about that (if people use it wrong, that's an all another story, if we start on that we might as well not have any filters).
As a matter of fact, I go further and I think that the core "Local Pickup" should have a specific field for pickup address (because the pickup location needs to be known by the customer) that should be injected into the area left blank by the not needed shipping address.

**Comment by mikejolley** [comment]
What about the filter on the shipping address itself?

**Comment by webdados** [comment]
Taking into consideration the argument that the order DOES needs shipping, what about a simple filter to override the shipping address / information?

Something like this:

On order-details-customer.php
```
<div class="woocommerce-column woocommerce-column--2 woocommerce-column--shipping-address col-2">
	<h2 class="woocommerce-column__title"><?php echo apply_filters( 'woocommerce_order_shipping_address_title', __( 'Shipping address', 'woocommerce' ) ); ?></h2>
	<address>
		<?php echo apply_filters( 'woocommerce_order_shipping_address_details', wp_kses_post( $order->get_formatted_shipping_address( __( 'N/A', 'woocommerce' ) ) ) ); ?>
	</address>
</div><!-- /.col-2 -->
```

On email-addresses.php:
```
<td style="text-align:<?php echo $text_align; ?>; font-family: 'Helvetica Neue', Helvetica, Roboto, Arial, sans-serif; padding:0;" valign="top" width="50%">
	<h2><?php echo apply_filters( 'woocommerce_order_shipping_address_title', __( 'Shipping address', 'woocommerce' ) ); ?></h2>
	<?php echo apply_filters( 'woocommerce_order_shipping_address_details', '<address class="address">'.$shipping.'</address>' ); ?>
</td>
```

On class-wc-admin-list-table-orders.php we would need another approach because it combines PHP a template and JS, but it's doable also.

**Comment by webdados** [comment]
`What about the filter on the shipping address itself?`

I was just writing a reply about that :-D

**Comment by webdados** [comment]
We cannot use the `woocommerce_order_formatted_shipping_address` filter because it works on an array, not on the final formatted address itself. We need to filter the final HTML.

**Comment by webdados** [comment]
For anyone interested, this was partially fixed on WooCommerce 8.6.0 by [adding the `woocommerce_email_customer_address_section` actions](https://github.com/woocommerce/woocommerce/blob/8.6.0/plugins/woocommerce/templates/emails/email-addresses.php).

The workaround I used to replace the shipping address when delivering to a parcelshop was something like this:
```
<?php
// Make sure we are on emails
add_action( 'woocommerce_email_order_details', function ( $order ) {
	if ( whatever logic to check if the order is delivered to a parcelshop ) {
		if ( version_compare( WC_VERSION, '8.6.0', '>=' ) ) {
			// Change the shipping address so it doesn't show the real one and instead show "Delivery in a parcelshop"
			// We cannot return false to the woocommerce_order_needs_shipping_address filter or else the section will not show at all
			add_filter( 'woocommerce_order_formatted_shipping_address', function ( $raw_address ) {
				return array(
					'first_name' => 'Delivery in a parcelshop',
				);
			} );
			add_action( 'woocommerce_email_customer_address_section', 'our_parcelshop_information' );
		} else {
			// Before WC 8.6, remove the need to show shipping addres and we'll just add the information after the addresses
			add_filter( 'woocommerce_order_needs_shipping_address', '__return_false' );
			add_action( 'woocommerce_email_customer_details', 'our_parcelshop_information' );
		}
	}
}
```
